dvc/remote/base.py

```diff
@@ -64,6 +66,24 @@ class RemoteMissingDepsError(DvcException):
     pass
 
 
+class TransferContextPool(object):
+    def __init__(self, func, stack):
+        self.stack = stack
+        self.func = func
+        self.pool = {}
+
+    def get(self):
+        tid = threading.current_thread().ident
+
+        ctx = self.pool.get(tid)
+        if not ctx:
+            ctx = self.func()
+            self.pool[tid] = ctx
```

Comment:
I would _double_ check that changing a dict this way is thread safe

Reply:
Built-in types are thread safe for single operations and we are doing this in one single thread(e.g. `tid` is unique to this thread, so we won't get a race condition in between pool.get() and pool[tid] set), so we should be safe. But we considering https://github.com/iterative/dvc/pull/2157#discussion_r295401140 , we could turn this thing around and create as many connections as we need and then create needed amount of workers.